server/tso/global_allocator.go

```diff
@@ -105,66 +147,111 @@ func (gta *GlobalTSOAllocator) SetTSO(tso uint64) error {
+	// (without synchronization with other Local TSO Allocators)
 	if len(dcLocationMap) == 0 {
 		return gta.timestampOracle.getTS(gta.leadership, count, 0)
 	}
 
 	// Have dc-locations configured in the cluster, use the Global TSO generation way.
-	// Send maxTS to all Local TSO Allocator leaders to prewrite
+	// (whit synchronization with other Local TSO Allocators)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	maxTSO := &pdpb.Timestamp{}
-	// 1. Collect the MaxTS with all Local TSO Allocator leaders first
-	if err := gta.SyncMaxTS(ctx, dcLocationMap, maxTSO); err != nil {
-		return pdpb.Timestamp{}, err
-	}
-	// 2. Add the count and make sure its logical part won't overflow after being differentiated.
-	suffixBits := gta.allocatorManager.GetSuffixBits()
-	gta.preprocessLogical(maxTSO, count, suffixBits)
-	// 3. Sync the MaxTS with all Local TSO Allocator leaders then
-	if err := gta.SyncMaxTS(ctx, dcLocationMap, maxTSO); err != nil {
-		return pdpb.Timestamp{}, err
-	}
-	// 4. Persist MaxTS into memory, and etcd if needed
+
 	var (
-		currentGlobalTSO pdpb.Timestamp
-		err              error
+		resp                   pdpb.Timestamp  = pdpb.Timestamp{}
+		estimatedMaxTSO        *pdpb.Timestamp = &pdpb.Timestamp{}
+		suffixBits             int             = gta.allocatorManager.GetSuffixBits()
+		shouldRetry, skipCheck bool
+		err                    error
 	)
-	if currentGlobalTSO, err = gta.getCurrentTSO(); err != nil {
-		return pdpb.Timestamp{}, err
-	}
-	if tsoutil.CompareTimestamp(&currentGlobalTSO, maxTSO) < 0 {
-		// Update the Global TSO in memory
-		if err := gta.timestampOracle.resetUserTimestamp(gta.leadership, tsoutil.GenerateTS(maxTSO), true); err != nil {
-			log.Warn("update the global tso in memory failed", errs.ZapError(err))
+	for i := 0; i < maxRetryCount; i++ {
+		// 1. Estimate a MaxTS among all Local TSO Allocator leaders according to the RTT if enableGlobalTSOEstimation.
+		if enableGlobalTSOEstimation {
+			estimatedMaxTSO, shouldRetry, err = gta.estimateMaxTS(count, suffixBits)
+			if err != nil {
+				return pdpb.Timestamp{}, err
+			}
+			if shouldRetry {
+				time.Sleep(gta.timestampOracle.updatePhysicalInterval)
+				continue
+			}
+		}
+	SETTING_PHASE:
+		// 2. Send the MaxTSO to all Local TSO Allocators leaders to make sure the subsequent Local TSOs will be bigger than it.
+		// It's not safe to skip check at the first time here because the estimated maxTSO may not be big enough,
+		// we need to validate it first before we write it into every Local TSO Allocator's memory.
+		resp = *estimatedMaxTSO
+		if err = gta.SyncMaxTS(ctx, dcLocationMap, &resp, skipCheck); err != nil {
+			return pdpb.Timestamp{}, err
+		}
+		// 3. If skipCheck is false and the maxTSO is bigger than estimatedMaxTSO,
+		// we need to redo the setting phase with the bigger one and skip the check safely.
+		if !skipCheck && tsoutil.CompareTimestamp(&resp, estimatedMaxTSO) > 0 {
```

Comment:
It's impossible `resp` is less than `estimatedMaxTSO` here. Because we have initialized `resp` as `estimatedMaxTSO` before and called `SyncMaxTS`. Either `resp` is as large as `estimatedMaxTSO` (a successful estimation), or `resp` becomes larger (a Local TSO bigger than `estimatedMaxTSO` is found during the synchronization).